### ISSUE TITLE
~~`bus error` on large datasets~~ Don't allocate big buffers on the stack

### ISSUE THREAD

**Issue by derekchiang** [ahrefs/ocannl#327]
I've been working on [a training script](https://github.com/derekchiang/ocannl/blob/bigrams/bin/bigram.ml) for a [bigram model](https://www.youtube.com/watch?v=PaCmpygFfXo) as a way to learn OCANNL.  It's working with small batch sizes like `100`, but when I try to [set the batch size](https://github.com/derekchiang/ocannl/blob/07fc4567722253f7340dda3abdf66ac35d86801c/bin/bigram.ml#L93) to something like 1000, I'm getting a `bus error`.  Is that expected?  If so, how do I train on a larger dataset?

The complete error message is:

```
[1]    32118 bus error  dune exec bin/bigram.exe
```

**Comment by derekchiang** [comment]
Can confirm that the same error persists with the latest `master` code.  I just [updated the script](https://github.com/derekchiang/ocannl/blob/bigrams/bin/bigram.ml) to merge with `master` in case you want to reproduce it.

I'm running on Macbook Air with Apple M2.

**Comment by lukstafi** [comment]
I'll look into this as soon as I fix a parallel training regression, then I'll release 0.6.0~beta -- before working on the new inline parameter definition syntax. Let's merge a PR for https://github.com/derekchiang/ocannl/tree/bigrams right after that. Not sure if it requires `uniform1` (the shape size divisibility issue with `uniform`).

**Comment by derekchiang** [comment]
Thank you sir -- this is indeed blocking me.  And yes I'd love to submit a PR for the bigrams example once we get it working.

**Comment by lukstafi** [comment]
Note that the sync_cc and multicore_cc backends (CPU) are currently more efficient than Metal and CUDA, and this will only change with OCANNL version 0.7. 

(Also, multicore_cc is way slower than expected, I was getting some speedups from parallelism while it was still buggy but once I removed buggy asynchronous updates it's only adding overhead... Definitely room for improvement but I'm not interested in optimizing it currently.)

**Comment by lukstafi** [comment]
This is fixed now, the error was due to stack overflow in the Domains runner.

**Comment by lukstafi** [comment]
Ah I've been blind, Claude's "fix" to this problem is only masking a nasty bug underneath. The current bigram test is all broken. It's assigning or treating as local memory mode to tensor nodes with data.

**Comment by lukstafi** [comment]
Not so nasty, I got confused.
